Cortex-M: Add depthwise conv2d operator

rascani opened the pull request:
### Summary
Add quantized depthwise convolution operator for the Cortex-M backend using CMSIS-NN's optimized arm_depthwise_conv_wrapper_s8 function.

Fixes #16105

### Test plan
```
./backends/cortex_m/test/build_test_runner.sh
pytest --config-file=backends/arm/test/pytest.ini backends/cortex_m/test/ops/test_conv.py
```

Commit:
Cortex-M backend: Add depthwise conv2d operator

Add quantized depthwise convolution operator for the Cortex-M backend
using CMSIS-NN's optimized arm_depthwise_conv_wrapper_s8 function.

Key changes:
- New op_quantized_depthwise_conv2d.cpp with CMSIS-NN implementation
- Python operator registration in operators.py with reference implementation
- Operator schema definition in operators.yaml
- Updated ConvertToCortexMPass to automatically detect and route depthwise
  convolutions (where groups == input_channels) to the specialized operator
- Comprehensive test coverage with 5 test cases covering different
  depthwise convolution scenarios (stride, padding, bias, depth multiplier)

The implementation validates the depthwise constraint (groups must equal
input channels) and supports NHWC layout, int8 quantization, per-channel
requantization, and configurable stride/padding/dilation parameters.

Commit:
Fix formatting

Commit:
Improve depthwise conv2d implementation with AOT optimizations and validations

Key changes:
- Move depth_multiplier calculation from runtime to AOT pass (eliminates
  runtime division by computing depth_multiplier = output_channels / input_channels
  in the graph transformation pass)
- Add critical defensive validations in validate_depthwise_conv2d_arguments():
  * Validate IHWO weight layout (dimension 0 must be 1)
  * Validate dilation == 1 (CMSIS-NN constraint)
  * Validate depth_multiplier consistency with channel counts
- Fix CMSIS-NN API usage:
  * Use arm_depthwise_conv_wrapper_s8_get_buffer_size() with correct parameters
  * Improve buffer allocation error handling with detailed error messages
- Add _compute_depthwise_conv2d_output_shape() to read channels from correct
  dimension (dim 3 for IHWO layout vs dim 0 for OHWI)
- Update operator schema to use depth_multiplier parameter instead of groups

This ensures proper validation of CMSIS-NN constraints and moves computation
to compile-time where possible.

Commit:
Add batch size validation and test coverage for depthwise conv2d

CMSIS-NN arm_depthwise_conv_wrapper_s8 only supports batch size 1.
Add validation in both AOT pass (fail during compilation) and runtime
(defensive check).

Add 6 test cases covering edge cases:
- Combined stride/padding/bias
- 1x1 kernels (common in mobile networks)
- Higher depth_multiplier (4)
- Asymmetric kernels (1x3)
- Asymmetric stride/padding
- Larger kernels (5x5)

Fix depthwise_conv2d_stride test to use batch size 1.

Commit:
Use is_channels_last_tensor helper function

Commit:
Remove invalid dilation check

Commit:
Shorten error messages

mansnils commented on backends/cortex_m/passes/convert_to_cortex_m_pass.py:
```diff
@@ -156,11 +156,24 @@ def _get_convolution_replacement(self, node) -> int:
+        # Detect depthwise convolution:
+        # PyTorch depthwise weight is [out_ch, 1, H, W] where dimension 1 is 1
+        # and groups == input_channels (groups > 1)
+        is_depthwise = weight_tensor.shape[1] == 1 and groups > 1
```
I think groups could be 1 for a DW conv?
So a better condition is then, is_depthwise = (in_channels == groups) and (out_channels % in_channels) == 0

rascani replied:
Done.

mansnils reviewed (changes requested):
Thanks for this @rascani ! It looks good, just a couple of comments.

psiddh commented on backends/cortex_m/ops/op_quantized_depthwise_conv2d.cpp:
```diff
@@ -0,0 +1,274 @@
+  const size_t buffer_bytes =
+      static_cast<size_t>(arm_depthwise_conv_wrapper_s8_get_buffer_size(
+          &dw_conv_params, &input_dims, &filter_dims, &output_dims));
+  if (buffer_bytes > 0) {
```
If the return val of  'arm_depthwise_conv_wrapper_s8_get_buffer_size' returns 'signed int' , should we add an  explicit '< 0'  check before we cast to size_t, because your current > 0 check happens after the potentially dangerous cast from signed to unsigned. (Just thinking aloud here..)

const int32_t raw_buffer_size = arm_depthwise_conv_wrapper_s8_get_buffer_size(
    &dw_conv_params, &input_dims, &filter_dims, &output_dims);
if (raw_buffer_size < 0) {
    ET_LOG(Error, "quantized_depthwise_conv2d_out: CMSIS-NN buffer size calculation failed");
    context.fail(Error::Internal);
    return out;
}

rascani commented on backends/cortex_m/passes/convert_to_cortex_m_pass.py:
```diff
@@ -156,11 +156,24 @@ def _get_convolution_replacement(self, node) -> int:
+        # and groups == input_channels (groups > 1)
+        is_depthwise = weight_tensor.shape[1] == 1 and groups > 1
+
+        if is_depthwise:
```
Good idea. I tried to add these constraints, which resulted in falling back to conv2d for some of the tests cases. Unfortunately, that also resulted in some output mismatches. I filed #16347 to revisit this and left a TODO to improve the smarts here.

Commit:
Fix depthwise conv routing and batch size handling

Two fixes for Cortex-M convolution pass:

1. Fix depthwise detection: Check `groups == in_channels` instead of
   `weight_tensor.shape[1] == 1 and groups > 1`. Previous logic failed
   for Conv2d(1, 4, 3) where in_channels=groups=1.

2. Route to standard conv when batch_size > 1: CMSIS-NN's
   `arm_depthwise_conv_wrapper_s8` falls back to unoptimized
   implementation for batch_size != 1, so check upfront and use
   standard conv instead of throwing ValueError.

Added `_get_batch_size_from_conv` helper that safely extracts batch
size from convolution node's output shape metadata. More reliable than
accessing input node metadata, which may be unavailable during
multi-pass transformations.

Fixes:
- conv2d_dilation: batch_size=3 with depthwise conv routes to standard conv
- conv2d_x3: KeyError accessing input metadata in multi-conv chains

rascani reviewed (commented):
Thanks for the reviews @mansnils and @psiddh!

psiddh reviewed (approved):
lgtm

Commit:
Merge remote-tracking branch 'upstream/main' into cmsis_depthwise_conv

Commit:
Fix post merge linting

zingo commented:
Hi this seem to make a cortex-m CI tests fails see https://hud.pytorch.org/hud/pytorch/executorch/main/1?per_page=50&name_filter=(arm-backend%7Ccortex)&useRegexFilter=true
